AWS Glue catalog and table operations

jackye1995 opened the pull request:
- Add `GlueCatalog` and `GlueTableOperations` implementations
- Fix bugs found in `S3FileIO` and add more tests
- add `:integrationTest` task to run integration tests for the `aws` module.

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+  private final String warehousePath;
+
+  /**
+   * constructor with default glue client
```
nit: Javadoc style. Capitalization and such

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+  }
+
+  /**
+   * constructor
```
nit: Javadoc style. Capitalization and probably needs elaboration

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+
+  private void validateWarehousePath() {
+    Preconditions.checkArgument(warehousePath != null && warehousePath.length() > 0,
+        "Glue warehouse path must be set");
```
nit: Iceberg error messages are usually "Cannot x because y. (Workaround or fix)"

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+  }
+
+  /**
+   * Use the same format as Hive for default warehouse location
```
This java doc isn't clear to me. I think the summary here should describe what the method is doing "Returns the default warehouse location as stored in the glue service (?). Uses the same format as Hive" 

Then under return "default warehouse path as a string using Hive's formatting"

Does that sound right?

jackye1995 replied:
Yeah let me just be precise with the behavior of it instead of saying it's the same as Hive.

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+    // should be safe to list all before returning the list, instead of dynamically load the list.
+    String nextToken = null;
+    List<TableIdentifier> results = new ArrayList<>();
+    while (true) {
```
I tend to be afraid of while(true) break, I would prefer this was written in a different style. One quick change here which wouldn't be that different would be a "do while" which would let you keep basically everything the same but you could drop the "break".

Ideally I like to avoid "break" if at all possible and especially in loops, but ymmv.

jackye1995 replied:
Sure, will use a do...while syntax for this.

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+    try {
+      dropTable(newTableOps(identifier), identifier, purge);
+      return true;
+    } catch (Exception e) {
```
Is there any possibility we might want to log the contents of this Exception? Feel's like we may want to surface some classes of errors.

kbendick replied:
Yeah I'd definitely like to log the contents of this exception. It would greatly help debug issues like whether or not the table drop failed due to invalid permissions, possibly the wrong role is being assumed at this point in time, etc.

jackye1995 replied:
Yes I definitely forgot to log here. Will be more detailed on this part.

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+    String nextToken = null;
+    List<Namespace> results = new ArrayList<>();
+    while (true) {
+      GetDatabasesResponse response = glue.getDatabases(GetDatabasesRequest.builder()
```
another maybe "do while" oppertunity

kbendick commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+    this.conf = conf;
+    this.glue = glue;
+    this.catalogId = catalogId;
+    this.warehousePath = warehousePath;
```
Nit: Can we add `Preconditions.checkNotNull` on anything that is nullable here so we can avoid nulll checks later?

jackye1995 replied:
warehousePath is checked by `validateWarehousePath` in the constructor, and `catalogId` can be null. I can add the precondition for `conf` and `glue`.

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+   * and DeleteTable or BatchDeleteTable, to delete any resources that belong to the database.
+   * @param namespace a namespace. {@link Namespace}
+   * @return always true
+   * @throws NamespaceNotEmptyException never thrown
```
I'm a little nervous about this implementation not following the pattern of the other catalogs and refusing to drop if there is something in the namespace. Downstream code may be relying on this to prevent destructive operations ...

jackye1995 replied:
Good point, this is definitely a different behavior between Glue and Hive. I can do a ListTable before dropping.

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,218 @@
+  private FileIO fileIO;
+
+  /**
+   * constructor
```
Nit: Java Doc style and summary need elaboration

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,218 @@
+  }
+
+  /**
+   * glue.catalogId.databaseName.tableName
```
This could just be a one liner

/**
* Returns table name in the form glue.catalogId.databaseName.tableName
**/

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,218 @@
+      persistGlueTable(isUpdate, parameters);
+      exceptionThrown = false;
+    } catch (ConcurrentModificationException e) {
+      throw new CommitFailedException(e, "Glue detected concurrent update to table %s", tableName());
```
More error messages which should be in the form Cannot x because y. I think the concurrent update to table error should also include the information 

"""   * For users that need high frequency and high concurrency update,
    * please consider requesting rate increase for Glue UpdateTable API,
    * or use an external lock system such as DynamoDB lock."""
Which an end user will usually not see

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueToIcebergConverter.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  /**
+   * validate the Glue table is Iceberg table by checking its parameters
```
Style on javadoc

RussellSpitzer commented on aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:
```diff
@@ -0,0 +1,66 @@
+  public static String toDatabaseName(Namespace namespace) {
+    if (!isValidNamespace(namespace)) {
+      throw new NoSuchNamespaceException(
+          "Bad namespace name %s, must be 1-232 chars of lowercase letters, numbers, underscore", namespace);
```
Exception message pattern

RussellSpitzer commented:
Looking good! Supporting more catalogs is great for the project!

yyanyy commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+  }
+
+  private TableMetadata dropTable(TableOperations ops, TableIdentifier identifier, boolean purge) {
+    TableMetadata lastMetadata = null;
+    if (ops.current() != null) {
+      lastMetadata = ops.current();
+    }
```
Nit: could be the same line

yyanyy commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    TableOperations ops = newTableOps(from);
+    // keep metadata
+    TableMetadata tableMetadata = dropTable(ops, from, false);
```
Do we want to validate `to` before dropping, so that we don't firstly drop table and then realize input was invalid?

jackye1995 replied:
good point, let me add the check

kbendick commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+          .catalogId(catalogId)
+          .databaseName(IcebergToGlueConverter.toDatabaseName(namespace))
+          .nextToken(nextToken)
+          .build());
```
You might need a `try ... catch` block here as the call to `getTables` can throw iirc.

jackye1995 replied:
I did not catch any exception because glue client only throw RuntimeException, and it seems to be okay in other catalog implementations.

rdblue replied:
It depends on the exception. If the namespace is missing, then we have an exception for that so users can handle it. We would want to translate the AWS exception to the Iceberg one. But for connection issues, it's okay to throw the original exception.

kbendick commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+      CatalogUtil.dropTableData(ops.io(), lastMetadata);
+    }
+
+    LOG.info("Dropped table: {}", identifier);
```
Nit: Maybe add an info log above when the table is dropped from the catalog and then one here about whether or not the table was purged?

kbendick commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+    glue.createDatabase(CreateDatabaseRequest.builder()
+        .catalogId(catalogId)
+        .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, metadata))
+        .build());
```
Seems like this operation can throw as well. Since the `BaseMetastoreCatalog` doesn't declare thi with a checked exception, you might want to document what it can throw that you know of or wrap in `try...catch` (though arguably that could be up the calling code).

jackye1995 replied:
Based on my observation of other catalogs, we are not handling runtime exceptions, and aws clients only throw runtime exception. That is why I do not try catch anything here. I can probably add a check for AlreadyExistsException.

kbendick commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,312 @@
+  public boolean setProperties(Namespace namespace, Map<String, String> properties) throws NoSuchNamespaceException {
+    Map<String, String> parameter = Maps.newHashMap();
+    parameter.putAll(loadNamespaceMetadata(namespace));
+    parameter.putAll(properties);
```
Maybe I'm just totally missing the obvious but I don't see where `parameter` is being used. Loks like it's built and then the database is updated directly via `propeties`.

jackye1995 replied:
It's used in `persistGlueTable` in `doCommit`

kbendick replied:
Ok. But to me, it seems like the locally declared `parameter` isn't used in this function.

Are you sure you didn't intend to use the merged `parameter` map instead of `properties` here when calling `IcebergToGlueConverter.toDatabaseInput(namespace, properties)`? https://github.com/apache/iceberg/pull/1633/files#diff-51a208168318cb652ad5beb86d3433d714aca916f9816c4cffcc88e273daf6afR287?

kbendick replied:
Conversely, the updated local metadata map is used in `removeProperties` instead of the passed in `properties`. So I still feel like potentially this variable is unused / was potentially meant to be passed into the `IcebergToGlueConverter.toDatabaseInput` instead of passing in the `properties` map.

Here's the usage in `removeProperties` that follows a similar pattern:
- Build an updated table properties map.
- Use the updated map when passing to the iceberg to glue converted for `databaseInput`

https://github.com/apache/iceberg/pull/1633/files#diff-51a208168318cb652ad5beb86d3433d714aca916f9816c4cffcc88e273daf6afR296-R302

jackye1995 replied:
Oh yeah sorry that one was actually my typo, thank you for pointing out. Somehow I was reading the wrong line when doing the reply...
I have pushed the latest code that has the fix and also with the integration test against actual Glue service that detected those bugs.

kbendick commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java:
```diff
@@ -0,0 +1,211 @@
+            Namespace.of("db2")
+        ),
+        glueCatalog.listNamespaces()
+    );
```
I don't really see much value in this test given how heavily mocked it is.

jackye1995 replied:
100% agree, I am thinking about also publish the actual Glue tests, but I have to also mark them as Ignored because it needs AWS credentials to run.

yyanyy commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalogInteg.java:
```diff
@@ -0,0 +1,271 @@
+    // Iceberg uses Hadoop 2.7.3, and the hadoop-aws module does not support reading environment variable yet
+    configuration.set("fs.s3a.access.key", System.getenv("AWS_ACCESS_KEY_ID"));
+    configuration.set("fs.s3a.secret.key", System.getenv("AWS_SECRET_ACCESS_KEY"));
+    glueCatalog = new GlueCatalog(configuration, GLUE, null, "s3a://" + BUCKET);
```
Not a problem for now but wondering if we want to add a path/folder with random name to the warehouse location inside bucket to allow concurrent runs

jackye1995 replied:
yeah that's a good idea, let me do that

yyanyy commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,351 @@
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    try {
+      TableOperations ops = newTableOps(identifier);
+      TableMetadata lastMetadata = ops.current() != null ? ops.current() : null;
```
Nit: could directly be `ops.current()`

yyanyy commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,351 @@
+    }
+    dropTable(from, false);
+    // use the same Glue info to create the new table, pointing to the old metadata
+    glue.createTable(CreateTableRequest.builder()
```
Sorry for coming back to this again, I wonder if it's worth adding a try catch block here and try recreating the old `fromTable` if this throws, so that for validation errors we didn't detect earlier or transient glue 5xx error we still could restore the table to the old state

jackye1995 replied:
sounds good, let me do that

yyanyy commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalogInteg.java:
```diff
@@ -0,0 +1,271 @@
+        IllegalArgumentException.class,
+        () -> glueCatalog.createTable(TableIdentifier.of(namespace, "table-1"), schema, partitionSpec));
+    // drop
+    glueCatalog.dropTable(TableIdentifier.of(namespace, tableName));
```
do we want to/is it easy to test the `purge` parameter?

jackye1995 replied:
yes I can add that test

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+    Preconditions.checkArgument(warehousePath != null && warehousePath.length() > 0,
+        "Cannot initialize GlueCatalog because warehousePath must not be null");
+    Preconditions.checkArgument(warehousePath.charAt(warehousePath.length() - 1) != '/',
+        "Cannot initialize GlueCatalog because warehousePath %s must not end with /", warehousePath);
```
Why not just remove the trailing `/`?

jackye1995 replied:
yeah that's actually better, let me do that

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+    if (dbLocationUri != null) {
+      return String.format("%s/%s", dbLocationUri, tableIdentifier.name());
+    }
+    return String.format(
```
Nit: we like to use an empty line after control flow statements to space out blocks a bit more.

rdblue replied:
Looks like this is needed in a lot of places.

jackye1995 replied:
sure, will update

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+    if (!namespace.isEmpty()) {
+      // if it is not a list all op, just check if the namespace exists and return itself.
+      if (namespaceExists(namespace)) {
+        return Lists.newArrayList(namespace);
```
Shouldn't this be an empty list? A namespace doesn't contain itself.

jackye1995 replied:
okay my misunderstanding, will fix

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+    }
+    // should be safe to list all before returning the list, instead of dynamically load the list.
+    String nextToken = null;
+    List<Namespace> results = new ArrayList<>();
```
Nit: we generally prefer factory methods like `Lists.newArrayList()` over specific list type constructors.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+          .catalogId(catalogId)
+          .nextToken(nextToken)
+          .build());
+      if (response.hasDatabaseList()) {
```
I think this needs an `else` to set `nextToken = null` and exit the loop. Otherwise, a response with no database list for a non-null token would cause an infinite loop because the token never changes.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+  @Override
+  public boolean dropNamespace(Namespace namespace) throws NamespaceNotEmptyException {
+    namespaceExists(namespace);
+    List<TableIdentifier> tableIdentifiers = listTables(namespace);
```
Does this need to check whether the namespace is empty?

jackye1995 replied:
yeah, that's why I try to list all tables of the namespace.

rdblue replied:
I mean whether the namespace is `Namespace.of()` and has no components. I think `loadNamespace` should handle it in the call to `toDatabaseName`, which ends up calling `checkNamespace`.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> properties) throws NoSuchNamespaceException {
+    Map<String, String> parameter = Maps.newHashMap();
```
This name isn't very descriptive. It seems to mean "a parameter that will be passed somewhere", but doesn't describe what the object is. I think a better name would be `newProperties` or similar. That fits the conventions we use elsewhere.

rdblue replied:
The method below uses `metadata`, which is also good.

jackye1995 replied:
Sure, `parameter` is used to mean `properties` in Glue that's why I used it. Will update to that.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,228 @@
+   * All-arg constructor
+   * @param conf Hadoop config, passed in for users of HadoopIO
+   * @param glue Glue client
+   * @param catalogId Glue catalog ID, which is the AWS account ID. When null, it uses the account of the Glue client.
```
Can Glue only have one catalog per account? Saying that the catalog ID "is" the AWS account ID seems to imply that.

Assuming that the account ID is a default catalog ID, I think that this should be more clear that the account ID is used by default, and that is the ID for the account's default catalog.

jackye1995 replied:
This is a very confusing concept within Glue... technically speaking each AWS account has one AWS Glue Data Catalog per AWS region. However, the `GlueClient` is always initialized with an AWS region, so the account ID is the only identifier needed to contact a different catalog. 

As a result, the `catalogId` in Glue always mean the AWS account ID, and that is what the user would like to configure outside the client. I can rewrite the doc to be more clear on this.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,228 @@
+    this.skipArchive = skipArchive;
+    this.databaseName = IcebergToGlueConverter.getDatabaseName(tableIdentifier);
+    this.tableName = IcebergToGlueConverter.getTableName(tableIdentifier);
+    this.fullName = String.format("glue.%s.%s.%s", catalogId, databaseName, tableName);
```
The full name here should be the name of the catalog, then the db and table name. This view is good for debugging (and may be what you want for `toString`) but the `name` returned for the table should be the name used to load it. So if I've called my Glue catalog `prodglue`, then it would be something like `prodglue.rdblue.test_table`.

jackye1995 replied:
I see. I am basically using `glue.<catalogId>` as the name, that is why I am writing it like this. Sounds like the `name` concept is used here differently, let me add this additional field then.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,228 @@
+          .databaseName(databaseName)
+          .tableInput(TableInput.builder()
+              .name(tableName)
+              .owner(System.getProperty("user.name"))
```
Is this required? Using `user.name` doesn't seem like a very accurate way to determine the user.

jackye1995 replied:
Agree, I am basically following what Hive is doing here. There is no other meaningful information that can be put as input here because the Glue client cannot provide the caller information. I can also just remove the field if that is cleaner.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,228 @@
+    } catch (OperationTimeoutException | ResourceNumberLimitExceededException e) {
+      throw new CommitFailedException(e,
+          "Cannot commit %s because Glue operation or resource limit exceeded, " +
+              "please consider contact AWS to raise the limit", tableName());
```
A timeout doesn't seem like the same case as exceeding a resource limit. Should these be handled separately?

jackye1995 replied:
Yeah I was expecting OperationTimeoutException to catch throttling due to API rate limit exceeded, but it also catches other situations. it is probably better to make it a different case.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,228 @@
+        isUpdate = true;
+        LOG.debug("Committing existing Glue table: {}", tableName());
+      } else {
+        LOG.debug("Committing new Glue table: {}", tableName());
```
I think it would make more sense to move these into `persistGlueTable` since that's when the actual call is made.

rdblue commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java:
```diff
@@ -0,0 +1,260 @@
+        .when(glue).getDatabase(Mockito.any(GetDatabaseRequest.class));
+    Mockito.doReturn(DeleteTableResponse.builder().build())
+        .when(glue).deleteTable(Mockito.any(DeleteTableRequest.class));
+    glueCatalog.dropTable(TableIdentifier.of("db1", "t1"));
```
This doesn't appear to verify that any of the above calls actually happened. Mock tests should validate that the expected RPC calls were actually made.

rdblue commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java:
```diff
@@ -0,0 +1,260 @@
+    );
+  }
+
+  @Test(expected = NoSuchNamespaceException.class)
```
It is usually better to use `AssertHelpers.assertThrows`. That way, you can check the error message and run additional assertions after the exception. For example, if you're testing create when a table already exists, you can validate that the right exception is thrown and that the existing table was not modified.

rdblue commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalogInteg.java:
```diff
@@ -0,0 +1,375 @@
+    glueCatalog.createNamespace(Namespace.of(namespace));
+    Database database = GLUE.getDatabase(GetDatabaseRequest.builder().name(namespace).build()).database();
+    Assert.assertEquals(namespace, database.name());
+    // list all
```
Can you separate the different calls into test methods? Combining a lot of cases together makes it harder to see what's wrong because failures prevent other test cases from running.

rdblue commented on build.gradle:
```diff
@@ -967,6 +967,24 @@ project(':iceberg-pig') {
+    compile project(':iceberg-bundled-guava')
+    compile project(':iceberg-core')
+    compile 'software.amazon.awssdk:glue'
+    compile 'software.amazon.awssdk:s3'
```
Why does this pull in S3? Is it needed?

jackye1995 replied:
I was thinking about the S3FileIO PR and marked it as compile dependency, let me change it for now to `testCompile`

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,398 @@
+   * Constructor that initializes all args from Hadoop configuration, used by engines like Spark and Flink
+   * @param conf Hadoop config passed from engines
+   */
+  public GlueCatalog(Configuration conf) {
```
I think this should use a config map like we're discussing in the catalog loading PR. We'll probably want to wait on this one until the direction there is agreed, and then update this.

rdblue commented:
Thanks, @jackye1995! This looks like a great start and mostly correct to me.

What is the plan for atomic commits? I don't see locking in this PR, so it looks like that will be added later with some configuration to enable it and point a catalog to a dynamo instance? It would also be great to know what the timeline for an atomic swap in Glue is. Maybe we could avoid adding the external locking, although it may be useful to have an external locking option.

jackye1995 commented:
> Thanks, @jackye1995! This looks like a great start and mostly correct to me.
> 
> What is the plan for atomic commits? I don't see locking in this PR, so it looks like that will be added later with some configuration to enable it and point a catalog to a dynamo instance? It would also be great to know what the timeline for an atomic swap in Glue is. Maybe we could avoid adding the external locking, although it may be useful to have an external locking option.

Currently the `doCommit` method uses Glue's `ConcurrentModificationException` to detect commit conflicts. However, because we are doing a client-side compare-and-swap, there is still a potential race condition. But the code should still work for low concurrency use cases. I am actively discussing with Glue about a native support, but it's not a very common use case for them to support yet at this point of time.

Yes, there will be another PR for adding the DynamoDB locking. When I add that, it will be enabled based on a config value, and user can choose to enable it or not based on the access pattern. As discussed in the original draft, I did not include this functionality to control the size of the PR. I already have the code, just waiting for this PR to be merged before adding that support.

jackye1995 commented:
Updated based on comments, and separated the integration tests into a separated gradle task, which can be executed by `gradle :iceberg-aws:integrationTest`. This can be executed in a different Github action and not affect a normal build.

rdblue commented:
@jackye1995, can you update this? #1573 was just merged. Thanks!

jackye1995 commented:
> @jackye1995, can you update this? #1573 was just merged. Thanks!

Sorry for the late reply, I was out for a few days. Just rebased the master, and also fixed some issues in `S3FileIO` found during testing.

rdblue commented:
> also fixed some issues in S3FileIO found during testing.

Are these blocking tests, or should we separate them into a different PR?

jackye1995 commented:
> > also fixed some issues in S3FileIO found during testing.
> 
> Are these blocking tests, or should we separate them into a different PR?

I am not entirely sure which way is the best to go here. Although there are 20 files changed, there are in total 6 main files changed and the rest are all tests so that's why I currently just put them all together. 

The blocking bug is in `S3OutputStream.close()` line 84, it needs to check `closed` boolean because `TableMetadataParser` tries to close the same input stream twice. ([link](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L117-L118)). All the other changes for S3 are just added tests.

I also can pull the fix and added tests for S3 out to another PR if that feels like a better way to go with smaller PRs, please let me know what you think.

jackye1995 commented:
@rdblue submitted PR #1749 for the `S3FileIO` bug fixes needed for this PR. I will also remove `S3FileIO` related integration tests.

rdblue commented on aws/src/integration/java/org/apache/iceberg/aws/glue/GlueCatalogNamespaceTest.java:
```diff
@@ -0,0 +1,127 @@
+  public void testCreateBadName() {
+    AssertHelpers.assertThrows("should not create namespace with bad name",
+        NoSuchNamespaceException.class,
+        () -> glueCatalog.createNamespace(Namespace.of("db-1")));
```
I don't think this should throw `NoSuchNamespaceException`. What about `ValidationException`?

Also, tests should generally check the error message as well.

rdblue replied:
Should this also test a namespace with more than 1 part?

jackye1995 replied:
Added. More details are directly tested in unit test of `IcebergToGlueConverter`

jackye1995 replied:
Yes `ValidationException` sounds better for the situations, I will update.

rdblue commented on aws/src/integration/java/org/apache/iceberg/aws/glue/GlueCatalogNamespaceTest.java:
```diff
@@ -0,0 +1,127 @@
+    namespaces.add(namespace);
+    glueCatalog.createNamespace(Namespace.of(namespace));
+    Database database = glue.getDatabase(GetDatabaseRequest.builder().name(namespace).build()).database();
+    Assert.assertEquals(namespace, database.name());
```
I usually like to check operations like `create` by checking the opposite state first, so asserting that the namespace does not exist using the glue client.

I also think it is a best practice to provide context in most assertions, like `assertEquals("Random name should match loaded when directly", namespace, database.name())`.

jackye1995 replied:
Will add messages to assertion.

rdblue commented on aws/src/integration/java/org/apache/iceberg/aws/glue/GlueCatalogTableTest.java:
```diff
@@ -0,0 +1,241 @@
+    s3.headObject(HeadObjectRequest.builder().bucket(testBucketName).key(key).build());
+    Table table = glueCatalog.loadTable(TableIdentifier.of(namespace, tableName));
+    Assert.assertEquals(partitionSpec, table.spec());
+    Assert.assertEquals(schema.toString(), table.schema().toString());
```
Can you add a bit more whitespace to break up sections of this test? It seems a bit dense.

rdblue commented on aws/src/integration/java/org/apache/iceberg/aws/glue/GlueCatalogTableTest.java:
```diff
@@ -0,0 +1,241 @@
+  @Test
+  public void testListTables() {
+    String namespace = createNamespace();
+    // list table should have nothing
```
This context is usually better if it is added to the assertion, so that when the test fails it is printed in the output.

rdblue commented on aws/src/integration/java/org/apache/iceberg/aws/glue/GlueCatalogTableTest.java:
```diff
@@ -0,0 +1,241 @@
+    AssertHelpers.assertThrows("should not have table",
+        NoSuchTableException.class,
+        () -> glueCatalog.renameTable(
+            TableIdentifier.of(namespace, tableName), TableIdentifier.of(namespace, tableName + "-3")));
```
What is a rename rollback? Why does this test fail? The table looks like it should exist.

jackye1995 replied:
Sorry the message is vague, and the test here is not very intuitive, I will update.

Because there is no actual `rename` operation in Glue, rename means deleting an old table and creating a new table in Glue pointing to the same metadata. If the creation process fails, it should "rollback" and keep the old table.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,423 @@
+   *                    So for streaming use case with lots of commits, it is recommended to set this value to true.
+   * @param fileIO file IO to use.
+   */
+  public GlueCatalog(GlueClient glue, String catalogName, String catalogId, String warehousePath,
```
Does this need to be public? It makes sense to have it for tests, but I would probably avoid public constructors that we need to support later.

rdblue replied:
Since all this does is call `initialize`, maybe we don't even need it for tests. That method could be package-private.

jackye1995 replied:
I will just have a package private constructor taking the glue client, and always call the initialize method then, which should simply the situation here a lot.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,423 @@
+        properties.get(CONFIG_GLUE_CATALOG_ID),
+        properties.get(CatalogProperties.WAREHOUSE_LOCATION),
+        skipArchiveValue == null ? CONFIG_SKIP_ARCHIVE_DEFAULT :
+            Boolean.parseBoolean(properties.get(CONFIG_SKIP_ARCHIVE)),
```
There's a helper for this: `PropertyUtil.propertyAsBoolean`.

jackye1995 replied:
Oh yes, completely forgot that, thank you

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,212 @@
+import software.amazon.awssdk.services.glue.model.TableInput;
+import software.amazon.awssdk.services.glue.model.UpdateTableRequest;
+
+public class GlueTableOperations extends BaseMetastoreTableOperations {
```
I don't think this class needs to be public or have a public constructor.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,400 @@
+        name,
+        properties.get(CatalogProperties.WAREHOUSE_LOCATION),
+        new AwsProperties(properties),
+        fileIOImpl == null ? new S3FileIO() : CatalogUtil.loadFileIO(fileIOImpl, properties, hadoopConf));
```
Not a blocker, but I wonder if we should choose the `FileIO` for a table based on the table location. Any s3 URI would use `S3FileIO`, but other URIs would still use the default `HadoopFileIO`.

Another thing I've been considering is using a delegating `FileIO` that checks whether `S3FileIO` is available and loads it. If it is available, then it is used for all s3 paths, but `HadoopFileIO` is used for the others. This may be over-complicated.

jackye1995 replied:
Yes I am also thinking about a delegating fileIO, but in another PR. I can switch default back to `HadoopFileIO` if that is more inclusive of all file path types.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,400 @@
+          .tableInput(tableInputBuilder.name(fromTableName).build())
+          .build());
+      throw e;
+    }
```
Seems like it would be safer to do this in the opposite order: create the new table as a copy and then drop the existing table. That way, if anything goes wrong, you have at least one copy of the table no matter what. And if only half of the operation succeeds, then you can clean it up by running DROP TABLE.

This way, if you lose connectivity after dropping the table, you may not be able to recover.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+    Table glueTable = null;
+    try {
+      Optional<Table> glueTableOptional = getGlueTable();
+      if (glueTableOptional.isPresent()) {
```
This logic is used in Hive because when running in Hive, `HiveMetaHook` is used to create the Iceberg table and by the time it runs the commit, Hive has already created the metadata in the HiveMetaStore.

This shouldn't be needed for Glue because we don't expect any other process to create the Glue table. Concurrent table creation should always call create, which may fail with an `AlreadyExistsException`.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+      throw e;
+    } catch (ConcurrentModificationException e) {
+      throw new CommitFailedException(e,
+          "Cannot commit %s because Glue detected concurrent update", tableName());
```
When would `ConcurrentModificationException` be thrown by Glue?

jackye1995 replied:
This is thrown when two processes call `glue.updateTable` at the same time, similar to `AlreadyExistsException` for `createTable`

rdblue replied:
I think this is the right way to handle that. I'm just curious about how this is caught. Is it a local check or a server-side check?

jackye1995 replied:
It's a server side check

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+          "Cannot commit %s because Glue detected concurrent update", tableName());
+    } catch (software.amazon.awssdk.services.glue.model.AlreadyExistsException e) {
+      throw new AlreadyExistsException(e,
+          "Cannot commit %s because its Glue table already exists when trying to create one", tableName());
```
How does this block know that `create` was called?

jackye1995 replied:
because `createTable` throws `AlreadyExistsException`, `updateTable` throws `ConcurrentModificationException`

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+        }
+      }
+
+      checkMetadataLocation(isUpdate, glueTable, base);
```
If the above block is removed, then this no longer needs to be in the `try` block. I think that also means you can remove `CommitFailedException` from the `catch`.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+          .databaseName(databaseName)
+          .name(tableName)
+          .build());
+      return Optional.ofNullable(response.table());
```
Why does this use `ofNullable`? Seems like the response returning a null table is an error.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+      persistGlueTable(isUpdate, properties);
+      exceptionThrown = false;
+    } catch (CommitFailedException | AlreadyExistsException e) {
+      throw e;
```
I don't think these exceptions are caught by other blocks, so you should be able to just remove this.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:
```diff
@@ -0,0 +1,198 @@
+    Map<String, String> properties = isUpdate ? Maps.newHashMap(glueTable.parameters()) : Maps.newHashMap();
+    properties.put(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(Locale.ENGLISH));
+    properties.put(METADATA_LOCATION_PROP, newMetadataLocation);
+    if (currentMetadataLocation() != null && !currentMetadataLocation().isEmpty()) {
```
If `currentMetadataLocation` is an empty string, then there is a problem. Empty strings shouldn't be handles like null strings.

rdblue replied:
That looks wrong to me, but if that's the currently released behavior then let's keep it. :frowning_face:

rdblue commented on aws/src/test/java/org/apache/iceberg/aws/glue/GlueCatalogTest.java:
```diff
@@ -0,0 +1,358 @@
+import software.amazon.awssdk.services.glue.model.UpdateDatabaseRequest;
+import software.amazon.awssdk.services.glue.model.UpdateDatabaseResponse;
+
+public class GlueCatalogTest {
```
These tests are okay, but it still doesn't seem like this is testing much. Most of the tests are responding to `Mockito.any(GetDatabaseRequest.class)` or similar, so it doesn't really test what gets passed to Glue. Better mocking would be great, but I'm starting to think that it may just be easier to maintain an in-memory catalog and a Glue client implementation that updates it. I think that would work better to catch simple problems.

rdblue commented:
Thanks @jackye1995! I think this is looking close. I had a few comments on the implementation in `GlueTableOperations` that I think should be considered.

Also, what is the plan for running the integration tests for this? I'm assuming that they pass when configured correctly. How can I run them and do you think we can make them part of CI?

jackye1995 commented:
@rdblue Thanks for the comments, I have been debating with myself about the rename order, and I have changed the dropTable to be the last operation. For the commit part, I have also simplified the logic based on the suggestions.

For the integration test part, one can run it by setting AWS credentials and a test bucket using environment variable `AWS_TEST_BUCKET` and run task `:iceberg-aws:integrationTest`. This should integrate well with Github actions. I am in the progress of requesting an account to run those tests for the open source community.

rdblue commented:
Merged! Thanks @jackye1995!
